pkg/kubelet/network/dns/dns.go

```diff
@@ -63,20 +66,50 @@ type Configurer struct {
 	// the container's DNS resolver configuration file. This can be used in
 	// conjunction with clusterDomain and clusterDNS.
 	ResolverConfig string
+
+	// client to access network CR
+	arktosV1Client arktosv1.ArktosV1Interface
 }
 
 // NewConfigurer returns a DNS configurer for launching pods.
-func NewConfigurer(recorder record.EventRecorder, nodeRef *v1.ObjectReference, nodeIP net.IP, clusterDNS []net.IP, clusterDomain, resolverConfig string) *Configurer {
+func NewConfigurer(recorder record.EventRecorder, nodeRef *v1.ObjectReference, nodeIP net.IP, clusterDNS []net.IP, clusterDomain, resolverConfig string, arktosV1Client arktosv1.ArktosV1Interface) *Configurer {
 	return &Configurer{
 		recorder:       recorder,
 		nodeRef:        nodeRef,
 		nodeIP:         nodeIP,
 		clusterDNS:     clusterDNS,
 		ClusterDomain:  clusterDomain,
 		ResolverConfig: resolverConfig,
+		arktosV1Client: arktosV1Client,
 	}
 }
 
+func (c *Configurer) getClusterDNS(pod *v1.Pod) []net.IP {
+	const defaultNetwork = "default"
+	networkName, ok := pod.Labels["arktos.futurewei.com/network"]
+	if !ok {
+		networkName = defaultNetwork
+	}
+	if network, err := c.arktosV1Client.NetworksWithMultiTenancy(pod.Tenant).Get(networkName, metav1.GetOptions{}); err == nil {
+		sip := network.Status.DNSServiceIP
+		ip := net.ParseIP(sip)
+		if ip != nil {
+			return []net.IP{ip}
+		}
+	} else {
+		klog.Errorf("failed to get network %s/%s: %v", pod.Tenant, networkName, err)
+	}
+
+	// for now, fallback to default cluster DNS for pod that has no explicit network and no valid dns service ip of the default network not set yet.
```

Comment:
Does a cluster still work if no network CRD is installed? I'm not sure if this is only temporary or it should be always there.

Reply:
I'm also thinking if it makes sense to make per-network optional. If we remove "if networkName == defaultNetwork" check here, we will have an effect of " if the network object has its own DNS, use it; otherwise fallback to the per-cluster one".  I'm still thinking about it -- not sure if it covers practical scenarios :)

Reply:
intention here is as temporary measure not to break system before the networking components are all ready (particularly the default network objects being created automatically on new tenant). I'll update code to make it clear.